sql-plugin/src/main/scala/com/nvidia/spark/rapids/RmmRapidsRetryIterator.scala

```diff
@@ -0,0 +1,393 @@
+   *
+   * When an attempt to invoke function `fn` is successful, the item K in `input` will be
+   * closed. In the case of a failure, all attempts will be closed. It is the responsibility
+   * of the caller to close any remaining items in `input` that have not been attempted.
+   *
+   * @tparam K element type that must be AutoCloseable
+   * @tparam T `fn` result type
+   * @param input an iterator of K
+   * @param fn a function that takes K and produces T
+   * @param splitPolicy a function that can split an item of type T into a Seq[T]. The split
+   *                    function must close the item passed to it.
+   */
+  class RmmRapidsRetryAutoCloseableIterator[K <: AutoCloseable, T](
+      input: Iterator[K],
+      fn: K => T,
+      splitPolicy: K => Seq[K])
+      extends RmmRapidsRetryIterator(input, fn, splitPolicy)
+        with Arm {
+
+    def this(
+        input: Iterator[K],
+        fn: K => T) = {
+      this(input, fn, null)
+    }
+
+    override def invokeFn(k: K): T = {
+      val res = super.invokeFn(k)
+      k.close() // close `k` only if we didn't throw from `invokeFn`
+      res
+    }
+
+    override def hasNext: Boolean = super.hasNext
+
+    override def next(): T = {
+      if (!hasNext) {
+        throw new IllegalStateException("Closed called on an empty iterator.")
+      }
+      try {
+        super.next()
+      } catch {
+        case t: Throwable =>
+          // exception occurred while trying to handle this retry
+          // we close our attempts (which includes the item we last attempted)
+          attemptStack.safeClose(t)
+          throw t
+      }
+    }
+  }
+
+  /**
+   * RmmRapidsRetryIterator exposes an iterator that can retry work,
+   * specified by `fn`, abstracting away the retry specifics.
+   *
+   * @tparam K element type that must be AutoCloseable
+   * @tparam T `fn` result type
+   * @param input an iterator of K
+   * @param fn a function that takes K and produces T
+   * @param splitPolicy an optional function that can split K into a Seq[K], if provided
+   *                    `splitPolicy` must take ownership of items passed to it.
+   */
+  class RmmRapidsRetryIterator[K, T](
+      input: Iterator[K],
+      fn: K => T,
+      splitPolicy: K => Seq[K]) extends Iterator[T] with Arm {
+    def this(input: Iterator[K], fn: K => T) =
+      this(input, fn, null)
+
+    protected val attemptStack = new mutable.ArrayStack[K]()
+
+    override def hasNext: Boolean = input.hasNext || attemptStack.nonEmpty
+
+    protected def invokeFn(k: K): T = {
+      fn(k)
+    }
+
+    override def next(): T = {
+      // this is set on the first exception, and we add suppressed if there are others
+      // during the retry attempts
+      var lastException: Throwable = null
+      var firstAttempt: Boolean = true
+      var result: Option[T] = None
+      var doSplit = false
+      while ((input.hasNext || attemptStack.nonEmpty) &&
```

Comment:
Can we switch the order of these? The input.hasNext could grab the GPU semaphore and actually materialize a batch on the assumption that it is going to be consumed soon.  Calling `hasNext` without calling next right away could cause memory leaks.

```
while (result.isEmpty && (attemptStack.nonEmpty || input.hasNext)) {
...
```

Reply:
nice catch